File: graph/l001.cpp
#include<iostream>
#include<vector>
#include<queue>
#include<algorithm>

using namespace std;

class Edge{
public:
    int v = 0;
    int w = 0;

    Edge(int v, int w){
        this->v = v;
        this->w = w;

    }
};

int N = 7;
vector<vector<Edge*>> graph(N,vector<Edge*>());

void addEdge(vector<vector<Edge*>> &gp, int u, int v, int w){

    gp[u].push_back(new Edge(v,w));
    gp[v].push_back(new Edge(u,w));

}

int findEdge(int v1 , int v2){

    int vtx=-1;
    for(int i=0; i<graph[v1].size(); i++){

        Edge* e = graph[v1][i];
        
        if(e->v == v2){
            vtx = i;
            break;
        }
    }
    return vtx;
}

void removeEdge(int u, int v){
    
    int idx1 = findEdge(u,v);
    int idx2 = findEdge(v,u);

    graph[u].erase(graph[u].begin() + idx1);
    graph[v].erase(graph[v].begin() + idx2);
}

void removeVtx(int vtx){

    while(graph[vtx].size()!=0){

        //Edge* e = graph[vtx][graph[vtx].size()-1];
        Edge* e = graph[vtx].back();
        removeEdge(vtx, e->v);
    }
}

void display(vector<vector<Edge*>> &gp){

    for(int i=0; i<N; i++){
        cout<<i<<"->";
        for(Edge* e: gp[i]){
            cout<<"("<<e->v<<","<<e->w<<")"<<",";
        }
        cout<<endl;
    }

    cout<<endl;
}

// Questions===================================

//flow of dfs -> 1.mark kro   then
//-------------> 2.nbrs nikalo vtx ke and call lgao.
bool hasPath(int src , int dest, vector<bool>& vis){

    if(src == dest){
        return true;
    }

    vis[src] = true;
    bool ans = false;
    for(int i=0; i<graph[src].size(); i++){
        Edge* e = graph[src][i];
        if(!vis[e->v]){
            ans = ans || hasPath(e->v , dest, vis);
        }
    }
    return ans;
}

int printPath(int src, int dest, int w, vector<bool>& vis, string ans){
    if(src == dest){
        cout<<ans+to_string(dest)<<"@"<<w<<endl;
        return 1;
    }

    vis[src] = true;
    int count = 0;
    for(int i=0; i<graph[src].size(); i++){
        Edge* e = graph[src][i];
        if(!vis[e->v]){
            count += printPath(e->v, dest, w + e->w, vis, ans + to_string(src));
        }
    }
    vis[src] = false;
    return count;
}

class AllsolutionPair{
public:
    int heavyW = 0; //largestpath
    string lps = "";
    int lightW = 1e7;//smallest path
    string sps = "";
    int ciel = 1e7;
    string cps = "";
    int floor = 0;
    string fps = "";
};
// find largest , smallest, ciel(smallest among larger than k), floor(largest among smaller than k) path 
//ciel - k se just bda path
//floor - k se just chota path.
int allSolution(int src, int dest, int w, vector<bool>& vis, string ans, AllsolutionPair &pair, int k){
    if(src == dest){
        //cout<<ans+to_string(dest)<<"@"<<w<<endl;
        if(w > pair.heavyW){
            pair.heavyW = w;
            pair.lps = ans + to_string(dest);
        }

        if(w < pair.lightW){
            pair.lightW = w;
            pair.sps = ans + to_string(dest);
        }

        if(w > k && w < pair.ciel){
            pair.ciel = w;
            pair.cps = ans + to_string(dest);
        }

        if(w < k && w > pair.floor){
            pair.floor = w;
            pair.fps = ans + to_string(dest);
        }

        return 1;
    }

    vis[src] = true;
    int count = 0;
    for(int i=0; i<graph[src].size(); i++){
        Edge* e = graph[src][i];
        if(!vis[e->v]){
            count += allSolution(e->v, dest, w + e->w, vis, ans + to_string(src), pair,k);
        }
    }
    vis[src] = false;
    return count;
}

int hamiltonianPath(int src, int osrc, vector<bool>& vis, int count, string ans){
    if(count == vis.size()-1){

        int idx = findEdge(src,osrc);
        if(idx != -1){
            cout<<"Cycle :"<<ans + to_string(src)<<endl;
        }else{
            cout<<"Path :"<<ans + to_string(src)<<endl;
        }
        return 1;
    }

    vis[src] = true;
    int cnt = 0;
    for(Edge* e: graph[src]){
        if(!vis[e->v]){
            cnt += hamiltonianPath(e->v, osrc, vis, count+1, ans + to_string(src));
        }
    }
    vis[src] = false;
    return cnt;
}

int dfs_components(int src , vector<bool> &vis){
    int count = 0;
    vis[src] = true;
    for(Edge* e:graph[src]){
        if(!vis[e->v])
            count += dfs_components(e->v, vis);
    }
    return count+1;
}

int getcomponents(){
    vector<bool> vis(N, false);

    int count = 0;
    int maxsize = 0;
    for(int i=0; i<N; i++){
        if(!vis[i]){
            count++;
            int cnt = dfs_components(i,vis);
            maxsize = max(cnt, maxsize);   
        }
    }
    //cout<<maxsize<<endl;
    return count;
}

string gscc_bfs(int src, vector<bool> &visited){

    queue<int> que;
    que.push(src);
    string comp ="";

    while(que.size()!=0){
        int rvtx = que.front(); que.pop();

        if(visited[rvtx]){
            continue;
        }
        visited[rvtx] = true;

        comp += to_string(rvtx);
        for(Edge* e: graph[rvtx]){
            if(!visited[e->v]){
                que.push(e->v);
            }
        }
    }
    return comp;
}

bool isConnected(){
    vector<bool> visited(N,false);
    int counter = 0;
    for(int i=0; i<N; i++){
        if(!visited[i]){
            //string ans = gscc_bfs(i,visited);
            int ans = dfs_components(i,visited);
            counter++;

            if(counter ==2)return false;
        }
    }
    return true;
}

vector<string> gcc(){
    vector<bool> visited(N,false);
    vector<string> ans;
    int counter = 0;
    for(int i=0; i<N; i++){
        if(!visited[i]){
            counter++;
            string comp = gscc_bfs(i,visited);
            ans.push_back(comp);
        }
    }
    cout<<counter<<endl;
    return ans;
}

bool isCyclic_bfs(int src, vector<bool>& vis){
    queue<int> que;
    que.push(src);
    //bfs ka 5th method lagaya h -> cycle vala
    while(que.size()!=0){
        int size = que.size();

        while(size-- >0){
            int rvtx = que.front(); que.pop();
            if(vis[rvtx]){
                return true;
            }
            vis[rvtx] = true;
            for(Edge* e: graph[rvtx])
                if(!vis[e->v]) que.push(e->v);
        }
    }
    return false;
}
bool isCyclic(){
    vector<bool> vis(N,false);
    bool ans = false;
    for(int i=0; i<N; i++){
        if(!vis[i]){
            ans = ans || isCyclic_bfs(i, vis);
        }
    }
    return ans;
}
//1st method where we use queue of type pair. pair stores two values int and string. 
void BFS_01(int src, vector<bool> &vis){
    queue<pair<int,string>> que;
    que.push({src, to_string(src) + ""});

    int dest = 6;

    while(que.size()!=0){

        pair<int,string> rvtx = que.front(); que.pop();

        if(vis[rvtx.first]){ //cycle hai?
            cout<<"Cycle: "<<rvtx.second<<endl;
            continue;
        }

        if(rvtx.first == dest){
            cout<<"destination: "<<rvtx.second<<endl;
        }

        vis[rvtx.first] = true;

        for(Edge* e: graph[rvtx.first]){

            if(!vis[e->v]){
                que.push({e->v, rvtx.second + to_string(e->v)});
            }
        }
    }
}

//2nd method where we make a pair class instead of using pair (usually c++ has pair but java doesnt, so make pair class in java)
void BFS_02(){
    //In java
    // delemeter vala h jisme null dalte h queue me 
    // aur jahan null hota h vahan pe level ka count bda lete.
}

void BFS_03(){
    //In java
    // delemeter vala h jisme null dalte h queue me 
    // yahan pair class me ek level name ka variable rkh lete aur usse bda lete jab bhi null mile.
}

void BFS_04(){
 // 2 QUEUE USE KRNE VALA METHOD.
}

void BFS_05(int src, vector<bool> &vis){

    queue<int> que;
    que.push(src);

    int level = 0;
    int cycle = 0;
    int dest = 6;

    while(que.size()!=0){
        int size = que.size();

        while(size-- >0){
            int rvtx = que.front(); que.pop();

            if(vis[rvtx]){ //cycle
                cout << "Cycle No. " + to_string(cycle) + ": " << rvtx << endl;
                cycle++;
                continue;
            }

            if(rvtx == dest){
                cout<<"dest:"<<level<<endl;
            }

            vis[rvtx] = true;
            for(Edge* e: graph[rvtx]){

                if(!vis[e->v]){
                    que.push(e->v);
                }
            }

        }
        level++;
    }
}

void BFS_06(int src, vector<bool> &vis){

    queue<int> que;
    que.push(src);
    vis[src] = true;

    int level = 0;
    int dest = 6;

    while(que.size() != 0){
        int size = que.size();

        while(size-- >0){
            int rvtx = que.front();
            que.pop();

            if(rvtx == dest){
                cout << "dest:" << level << endl;
            }

            for(Edge* e: graph[rvtx]){
                
                if(!vis[e->v]){

                    que.push(e->v);
                    vis[e->v] = true;
                }
            }

        }
        level++;
    }
}

bool isBipartiteBfs_02(int src, vector<int>& vis){
    
    queue<pair<int,int>> que; //int-src, int-level; 
    que.push({src,0});

    while(que.size()!=0){
        pair<int,int> vtx = que.front(); que.pop();

        if(vis[vtx.first] != 0){
            //do some work.
            if(vis[vtx.first]%2 != vtx.second%2)
                return false;
        }
        vis[vtx.first] = vtx.second;

        for(Edge* e: graph[vtx.first])
            if(vis[e->v] == 0)
                que.push({e->v, vtx.second + 1});
    }
    return true;
}

bool isBipartite_02(){
    vector<int> vis(N, 0);
    for(int i=0; i<N; i++){
        if(vis[i]==0){
            bool ans = isBipartiteBfs_02(i, vis);
            if(ans==false)
                return false;
        }
    }
    return true;
}

bool isBipartiteBfs(int src, vector<int> &vis){

    queue<pair<int,int>> que; //int -> src, int-> color(0,1,-1)
    que.push({src, 0});

    int cycle = 0;

    while(que.size()!=0){
        int size = que.size();
        while (size-- > 0)
        {
            pair<int, int> rvtx = que.front();
            que.pop();

            if (vis[rvtx.first] != -1) // already visited(cycle).
            {   cycle++;
                if (vis[rvtx.first] != rvtx.second) // check for conflict.
                    return false;
            }

            vis[rvtx.first] = rvtx.second;
            for (Edge* e : graph[rvtx.first])
            {
                if (vis[e->v] == -1)
                    que.push({e->v, (rvtx.second + 1) % 2}); 
            }
        }
    }
}
void isBipartite(){

    vector<int> vis(N,-1); // -1 ->unvisited 0 ->red 1 -> green
    for(int i=0; i<N; i++){
        if(vis[i]==-1){
            cout<<isBipartiteBfs(i,vis)<<endl;
        }
    }
}

vector<int> par;
vector<int> setSize;

int findPar(int vtx){
    if(par[vtx] == vtx)
        return vtx;
    return par[vtx] = findPar(par[vtx]); //Path Compression -> overall parent nikal rhe h.
}

void mergeSet(int l1, int l2){
    if(setSize[l1] < setSize[l2]){

        par[l1] = l2;
        setSize[l2] += setSize[l1];
    }
    else{
        par[l2] = l1;
        setSize[l1] += setSize[l2];
    }
}

// void kruskalsAlgo(vector<vector<int>>& arr){

//     vector<vector<Edge*>> Kgraph(arr.size(), vector<Edge*>());
    
//     for(vector<int> &ar: arr){

//         int u = arr[0];
//         int v = ar[1];

//         int p1 = findPar(u);
//         int p2 = findPar(v);

//         if(p1 != p2){
//             mergeSet(p1, p2);
//             addEdge()
//         }
//     }
// }

class dpair{
    public:
        int src;
        int parent;
        int w;
        int wsf;

        dpair(int src, int parent, int w, int wsf){
            this-> src = src;
            this->parent = parent;
            this->w= w;
            this->wsf = wsf;
        }
};

struct Dcomp{
    public:
        bool operator()(dpair &p1, dpair &p2){
            return p1.wsf > p2.wsf;  //default behavior (this - other) - max.
            // return p2.wsf > p1.wsf;
        }
};

void dijikstraAlgo(int src){

    vector<vector<Edge*>> dkgraph(N, vector<Edge*>());

    priority_queue<dpair, vector<dpair>, Dcomp> pq;
    vector<bool> vis(N, false);
    pq.push(dpair(src, -1, 0, 0));

    while(pq.size() !=0){
        int size = pq.size();
        while(size-- >0){
            dpair vtx = pq.top(); pq.pop();

            if(vis[vtx.src])
                continue;

            if(vtx.parent != -1)
                addEdge(dkgraph, vtx.src, vtx.parent, vtx.w);

            vis[vtx.src] = true;

            for(Edge* e: graph[vtx.src]){
                if(!vis[e->v])
                    pq.push(dpair(e->v, vtx.src, e->w, vtx.wsf + e->w));
            }
        }
    }

    display(dkgraph);
}

class ppair{
    public:
        int src;
        int parent;
        int w;

        ppair(int src, int parent, int w){
            this->src = src;
            this->parent  = parent;
            this->w = w;
        }
};

struct Pcomp{
    public:
        bool operator()(ppair &p1, ppair &p2){
            return p1.w > p2.w; 
        }
};

void PrimsAlgo(int src){

    vector<vector<Edge *>> Pgraph(N, vector<Edge *>());

    priority_queue<ppair, vector<ppair>, Pcomp> pq;
    pq.push(ppair(src, -1, 0));
    vector<bool> vis(N, false);

    while(pq.size() != 0){
        int size = pq.size();
        while(size-- >0){
            ppair rvtx = pq.top();
            pq.pop();

            if(vis[rvtx.src])
                continue;
            
            if(rvtx.parent != -1)
                addEdge(Pgraph, rvtx.src, rvtx.parent, rvtx.w);
            
            vis[rvtx.src] = true;

            for(Edge * e : graph[rvtx.src])
                if(!vis[e->v])
                    pq.push(ppair(e->v, rvtx.src, e->w));
        }
    }
    display(Pgraph);
}
void contsructgraph(){

    addEdge(graph,0,1,20);
    addEdge(graph,1,2,20);
    addEdge(graph,0,3,10);
    addEdge(graph,2,3,40);
    addEdge(graph,3,4,2);
    addEdge(graph,4,5,2);
    addEdge(graph,5,6,8);
    addEdge(graph,4,6,3);
    // addEdge(graph,2,5,2);

    display(graph);
}

void questions(){
    // removeEdge(3,4);
    // display(graph);
    // removeVtx(3);
    // display(graph);

    vector<bool> vis(N,false);
    //cout<<hasPath(0,6,vis)<<endl;
    //cout<<"--------"<<endl;
    //cout<<printPath(0,6,0,vis,"")<<endl;
    // cout<<"--------"<<endl;
    // AllsolutionPair pair;
    // cout<<allSolution(0,6,0,vis,"",pair,20)<<endl;
    // cout<<"Largest-:"<<pair.lps<<"@"<<pair.heavyW<<endl;
    // cout<<"smallest-:"<<pair.sps<<"@"<<pair.lightW<<endl;
    // cout<<"ciel path-:"<<pair.cps<<"@"<<pair.ciel<<endl;
    // cout<<"floor path-:"<<pair.fps<<"@"<<pair.floor<<endl;

    //cout<<hamiltonianPath(2,2,vis,0,"")<<endl;
    //cout<<getcomponents()<<endl;
    
    // BFS_05(0,vis);
    // BFS_04(0, vis);
    // cout<<(boolalpha)<<isConnected()<<endl;
    // vector<string> ans = gcc();
    // for(string val: ans){
    //     cout<<val<<" ";
    // }
    // cout<<endl;
    //cout<<(boolalpha)<<isBipartite_02()<<endl;

    dijikstraAlgo(2);
    PrimsAlgo(2);
    PrimsAlgo(6);
}

int main(){
    contsructgraph();
    questions();
    return 0;
}

File: graph/l001.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.PriorityQueue;

class l001{
    public static void main(String[] args){
        solve();
    }

    public static void solve(){
        constructgraph();
    }

    static class Edge{
        int v = 0;
        int w = 0;

        Edge(int v, int w){
            this.v = v;
            this.w = w;
        }
    }

    static int N = 7;
    static ArrayList<Edge>[] graph;

    public static void addEdge(ArrayList<Edge>[] gp, int u, int v, int w){

        gp[u].add(new Edge(v,w));
        gp[v].add(new Edge(u,w));
    }

    public static void display(ArrayList<Edge>[] gp){
        for(int i=0; i<N; i++){
            System.out.print(i+" -> ");
            for(Edge e: gp[i]){
                System.out.print("(" + e.v + "," + e.w + ")" + ", ");
            }
            System.out.println();
        }

        System.out.println();
    }

    public static int findEdge(int v1, int v2){

        int vtx = -1;
        for(int i=0; i<graph[v1].size(); i++){

            Edge e = graph[v1].get(i);
            if(e.v == v2){
                vtx = i;
                break;
            }
        }

        return vtx;
    }

    public static void removeEdge(int u, int v){

        int idx = findEdge(u, v);
        int idx2 = findEdge(v, u);

        graph[u].remove(idx);
        graph[v].remove(idx2);
    }

    public static void removeVtx(int vtx){

        while(graph[vtx].size()!=0){

            Edge e = graph[vtx].get(graph[vtx].size()-1);
            removeEdge(vtx, e.v);
        }
    }

    public static boolean hasPath(int src, int dest, boolean[] vis){

        if(src == dest) return true;

        vis[src] = true;
        boolean ans = false;
        for(Edge e: graph[src]){
            if(!vis[e.v]){
                ans = ans || hasPath(e.v, dest, vis);
            }
        }
        return ans;
    }

    public static int printPath(int src, int dest, int w, boolean[] vis, String ans){

        if(src == dest){
            System.out.println(ans + src + "@" + w);
            return 1;
        }

        vis[src] = true;
        int count = 0;
        for(Edge e: graph[src]){
            if(!vis[e.v]){
                count += printPath(e.v, dest, w + e.w, vis, ans + src + " ");
            }
        }
        vis[src] = false;
        return count;
    }

    static class Pair{
        int largeW = 0;
        String lps = "";
        int smallW = (int)1e7;
        String sps = "";
        int ciel = (int)1e7;
        String cps = "";
        int floor = 0;
        String fps = "";
    }

    public static void allSolution(int src, int dest, int w, boolean[] vis, String ans, Pair pair, int K){

        if(src == dest){
            // System.out.println(ans + src + "@" + w);
            //largest path
            if(w > pair.largeW){
                pair.largeW = w;
                pair.lps = ans + src;
            }
            // smallest path.
            if(w < pair.smallW){
                pair.smallW = w; 
                pair.sps = ans + src;
            }            
            //ciel path -> largest valon me se sabse chota.
            if(w > K && w < pair.ciel){
                pair.ciel = w;
                pair.cps = ans + src;
            }
            // floor path -> smallest valon me se sabse bda.
            if(w < K && w > pair.floor){
                pair.floor = w;
                pair.fps = ans + src;
            }
            return ;
        }

        vis[src] = true;
        for(Edge e: graph[src]){
            if(!vis[e.v]){
                allSolution(e.v, dest, w + e.w, vis, ans + src + " ",pair,K);
            }
        }
        vis[src] = false;
        return ;
    }

    // static class Pair{
    //     int vtx;
    //     String  psf;
    //     int level;

    //     Pair(int vtx, String psf){
    //         this.vtx = vtx;
    //         this.psf = psf;
    //     }

    //     Pair(int vtx, String psf, int level){
    //         this.vtx = vtx;
    //         this.psf = psf;
    //         this.level = level;
    //     }
    // }

    // static void BFS_01(int src, boolean[] vis){
    //     LinkedList<Pair> que = new LinkedList<>();
    //     que.addLast(new Pair(src,  src + ""));

    //     int level =0;
    //     int desti = 4;

    //     while(que.size()!=0){

    //         Pair rvtx = que.removeFirst();

    //         if(vis[rvtx.vtx]){
	// 			System.out.println("Cycle: " + rvtx.psf);
	// 			continue;
	// 		}

	// 		if(rvtx.vtx==desti){
	// 			System.out.println("destinantion: " + rvtx.psf);
	// 		}
			
	// 		vis[rvtx.vtx]=true;

    //         for(Edge e:graph[rvtx.vtx]){
    //             if(!vis[e.v]){
    //                 que.addLast(new Pair(e.v, rvtx.psf + e.v));
    //             }
    //         }
    //     }
    // }

    // static void BFS_02(int src, boolean[] vis){
    //     LinkedList<Pair> que = new LinkedList<>();
    //     que.addLast(new Pair(src, ""));
    //     que.addLast(null);

    //     int level =0;
    //     int dest = 6;

    //     while(que.size()!=0){

    //         Pair rvtx = que.removeFirst();

    //         if(vis[rvtx.vtx]){
	// 			System.out.println("Cycle: " + rvtx.psf);
	// 			continue;
	// 		}

	// 		if(rvtx.vtx==desti){
	// 			System.out.println("destinantion: " + rvtx.psf + " -> " + level);
	// 		}
			
	// 		vis[rvtx.vtx]=true;

    //         for(Edge e:graph[rvtx.vtx]){
    //             if(!vis[e.v]){
    //                 que.addLast(new Pair(e.v, rvtx.psf));
    //             }
    //         }

    //         if(que.getFirst()==null){
	// 			level++;
	// 			que.removeFirst();
	// 			que.addLast(null);
	// 		}
    //     }
    // }

    // static void BSF_03(int src, boolean[] vis){

    //     LinkedList<Pair> que = new LinkedList<>();
    //     que.addLast(new Pair(src, src+" "));
    //     que.addLast(null);

    //     int dest = 6;
        
    //     while(que.size()!=0){
    //         Pair rvtx = que.removeFirst();

    //         if(vis[rvtx.vtx]){
    //             System.out.println("Cycle: "+ rvtx.psf);
    //             continue;
    //         }

    //         if(rvtx.vtx == dest){
    //             System.out.println("destination: " + rvtx.psf + "->" + rvtx.level);
    //         }

    //         vis[rvtx.vtx] = true;

    //         for(Edge e: graph[rvtx.vtx]){
    //             if(!vis[rvtx.vtx])
    //                 que.addLast(new Pair(e.v, rvtx.psf + e.v, rvtx.level+1));
    //         }

    //     }
    // }

    // static class Pair{
    //     int s;
    //     int color;
        
    //     Pair(int src, int color){
    //         this.s = src;
    //         this.color = color;
    //     }
    // }
    
    // public static boolean isBipartite_bfs(int[][] graph, int src, int[] vis){
        
    //     LinkedList<Pair> que = new LinkedList<>();
    //     que.addLast(new Pair(src, 0));
        
    //     while(que.size()!=0){
    //         int size = que.size();
            
    //         while(size-- >0){
    //             Pair vtx = que.removeFirst();
                
    //             if(vis[vtx.s] != 0){
                    
    //                 if(vis[vtx.s]%2 != vtx.color%2)
    //                     return false;
    //             }
                
    //             vis[vtx.s] = vtx.color;
                
    //             // int i = vtx.s/graph[0].length;
    //             // int j = vtx.s%graph[0].length;
                
    //             for(int e : graph[vtx.s]){
    //                 // int nbr = e * graph[0].length + j;
    //                 if(vis[e] == 0)
    //                     que.addLast(new Pair(e, vtx.color + 1));
    //             }
    //         }
    //     }
    //     return true;
    // }
    
    //another method - bipartite using int[]..
    public static boolean isBipartite_bfs_02(ArrayList<Edge>[] graph, int src, int[] vis){
        
        LinkedList<int[]> que = new LinkedList<>();
        que.addLast(new int[]{src, 0});
        
        while(que.size()!=0){
            int size = que.size();
            
            while(size-- >0){
                int[] vtx = que.removeFirst();
                
                if(vis[vtx[0]] != 0){
                    
                    if(vis[vtx[0]]%2 != vtx[1]%2)
                        return false;
                }
                
                vis[vtx[0]] = vtx[1];
                
                // int i = vtx.s/graph[0].length;
                // int j = vtx.s%graph[0].length;
                
                for(Edge e : graph[vtx[0]]){
                    // int nbr = e * graph[0].length + j;
                    if(vis[e.v] == 0)
                        que.addLast(new int[]{e.v, vtx[1] + 1});
                }
            }
        }
        return true;
    }

    public static boolean isBipartite(ArrayList<Edge>[] graph) {
        
        int n = graph.length;
        if(n == 0) return true;
        
        int m = graph[0].size();
        
        int[] vis = new int[n];
        
        for(int i=0; i<n; i++){
            if(vis[i] == 0){
                boolean ans = isBipartite_bfs_02(graph, i, vis);
                if(!ans)return false;
            }
        }
        return true;
    }

    // union-find
    // static int[] par;
    // static int[] setSize;

    // int findPar(int vtx){
        
    //     if(par[vtx] == vtx) return vtx;
    //     return par[vtx] = findPar(par[vtx]);
    // }

    // void mergeSet(int l1, int l2){
    //     if(setSize[l1] < setSize[l2]){
    //         par[l1] = l2;
    //         setSize[l2] += setSize[l1];
    //     }
    //     else{
    //         par[l2] = l1;
    //         setSize[l1] += setSize[l2];
    //     }
    // }
    // //now we can use this in our questions...
    // //leetcode-684
    // public int[] findRedundantConnection(int[][] edges) {
        
    //     int n = edges.length;
    //     par = new int[n];
    //     setSize = new int[n];

    //     for(int i=0; i<=n; i++) par[i] = i; setSize[i] = 1;

    //     for(int[] ar: edges){

    //         int p1 = findPar(ar[0]);
    //         int p2 = findPar(ar[1]);

    //         if(p1 != p2) mergeSet(p1, p2);
    //         else
    //             return ar;
    //     }

    //     return new int[]{};
    // }

    // //leetcode-547 - famous question hai.
    // public int findCircleNum(int[][] M) {
        
    //     if(M.length == 0) return 0;
        
    //     int n = M.length;
    //     int m = M[0].length;
    //     par = new int[n];

    //     for(int i=0; i<n; i++) par[i] = i;

    //     countFc = n;
    //     for(int i=0; i<n; i++){
    //         for(int j=0; j<m; j++){
    //             if(M[i][j] == 1 && i !=j){

    //                 int p1 = findPar(i);
    //                 int p2 = findPar(j);

    //                 if(p1 != p2){ 
    //                     par[p2] = p1;
    //                     countFc--;
    //                 }
    //             }
    //         }
    //     }
    //     return countFc;
    // }

    public static class dpair{
            int src;
            int parent;
            int w;
            int wsf;
    
            dpair(int src, int parent, int w, int wsf){
                this.src = src;
                this.parent = parent;
                this.w= w;
                this.wsf = wsf;
            }
    }
    
    static void dijikstraAlgo(int src){
        
        ArrayList<Edge>[] dkgraph = new ArrayList[N];
        for(int i=0; i<N; i++){
            dkgraph[i] = new ArrayList<Edge>();
        }

        PriorityQueue<dpair> pq = new PriorityQueue<>((dpair a, dpair b)->{
            return a.wsf - b.wsf; //min pq (by default);
            //return b.wsf - a.wsf ; //max pq
        });

        boolean[] vis = new boolean[N];
        pq.add(new dpair(src, -1, 0, 0));
    
        while(pq.size() !=0){
            int size = pq.size();
            while(size-- >0){
                dpair vtx = pq.poll();
                
                if(vis[vtx.src]){
                    continue;
                }

                if(vtx.parent != -1){
                    addEdge(dkgraph, vtx.src, vtx.parent, vtx.w);
                }
    
                vis[vtx.src] = true;
    
                for(Edge e: graph[vtx.src]){
                    if(!vis[e.v])
                        pq.add(new dpair(e.v, vtx.src, e.w, vtx.wsf + e.w));
                }
            }
        }

        display(dkgraph);
    }

    public static class ppair{
        int src;
        int parent;
        int w;

        ppair(int src, int parent, int w){
            this.src = src;
            this.parent = parent;
            this.w= w;
        }
}

    static void primsAlgo(int src){

        ArrayList<Edge>[] pGraph = new ArrayList[N];
        for(int i=0; i<N; i++){
            pGraph[i] = new ArrayList<Edge>();
        }
    
        PriorityQueue<ppair> pq = new PriorityQueue<>((ppair a, ppair b)->{
            return a.w - b.w; //min pq (by default);
            //return b.wsf - a.wsf ; //max pq
        });

        boolean[] vis = new boolean[N];
        pq.add(new ppair(src, -1, 0));
    
        while(pq.size() !=0){
            int size = pq.size();
            while(size-- >0){
                ppair vtx = pq.poll();
                
                if(vis[vtx.src]){
                    continue;
                }

                if(vtx.parent != -1){
                    addEdge(pGraph, vtx.src, vtx.parent, vtx.w);
                }
    
                vis[vtx.src] = true;
    
                for(Edge e: graph[vtx.src]){

                    if(!vis[e.v])
                        pq.add(new ppair(e.v, vtx.src, e.w));
                }
            }
        }

        display(pGraph);
    }
    
    public static void constructgraph(){

        graph = new ArrayList[N];
        for(int i=0; i<N; i++){
            graph[i] = new ArrayList<Edge>();
        }

        addEdge(graph, 0, 1, 20);
        addEdge(graph, 1, 2, 10);
        addEdge(graph, 2, 3, 40);
        addEdge(graph, 0, 3, 20);
        addEdge(graph, 3, 4, 2);
        addEdge(graph, 4, 5, 6);
        addEdge(graph, 4, 6, 8);
        addEdge(graph, 5, 6, 3);

        // display(graph);

        // removeEdge(3, 4);
        // display(graph);

        // removeVtx(3);
        // display(graph);

        boolean[] vis = new boolean[N];
        // BFS_01(0, vis);
        // System.out.println(hasPath(0, 6, vis));
        // System.out.println(printPath(0,6,0, vis,""));
        // System.out.println("---------------------");
        // Pair pair = new Pair();
        // allSolution(0,6,0, vis,"",pair,20);
        // System.out.println("Largest-:"+pair.lps+"@"+pair.largeW);
        // System.out.println("smallest-:"+pair.sps+"@"+pair.smallW);
        // System.out.println("ciel path-:"+pair.cps+"@"+pair.ciel);
        // System.out.println("floor path-:"+pair.fps+"@"+pair.floor);

        //System.out.println(isBipartite(graph));

        // primsAlgo(0);
        // primsAlgo(6);
        dijikstraAlgo(2);
        

    }
}